swev-id: django__django-12419 - Default SECURE_REFERRER_POLICY to 'same-origin' and document BC

rowan-stein opened the pull request:
This PR implements the secure-by-default change proposed in Issue #194: change the default for SECURE_REFERRER_POLICY to 'same-origin'.

Summary
- Code: Set SECURE_REFERRER_POLICY default to 'same-origin' in django/conf/global_settings.py.
- Docs:
  - Update docs/ref/settings.txt to reflect the new default and add a versionchanged note (3.1).
  - Add a Backwards incompatible changes note in docs/releases/3.1.txt with mitigation guidance.
  - Mention the default in docs/ref/middleware.txt.
- Tests: Existing tests continue to pass locally for middleware and checks that explicitly override settings. No changes required.

Backward compatibility
- SecurityMiddleware will now add a Referrer-Policy header by default ('same-origin').
- If your deployment relies on omitting or customizing this header, set SECURE_REFERRER_POLICY accordingly (e.g., None to omit, or 'no-referrer').
- Note: For HTTPS, Django’s CSRF protection requires the Referer header. The default 'same-origin' generally protects privacy while maintaining CSRF functionality.

Local verification
- Ran focused test modules locally:
  - tests/middleware/test_security.py
  - tests/check_framework/test_security.py
- Result: All tests passed.

Observed failure during local setup and resolution
- Initial test run failed because Python 3 was missing. Installed Python 3.10 via Nix.
- Attempt to use tox v4 failed due to config incompatibility. Switched to Django’s test runner directly.
- Missing dependencies caused ModuleNotFoundError (e.g., asgiref). Installed test requirements via pip. A build error occurred for pywatchman due to missing GCC; resolved by installing gcc via Nix and re-running dependency installation.
- After setup, targeted tests passed.

References
- Issue: https://github.com/agyn-sandbox/django/issues/194
- Ticket context: #29406 added SECURE_REFERRER_POLICY in Django 3.0.

Note
- Base branch: django__django-12419.
- CI: Not manually triggered.

rowan-stein committed:
Change default SECURE_REFERRER_POLICY to 'same-origin' (swev-id: django__django-12419). Add inline comment noting version change and how to omit header.

rowan-stein committed:
Docs: SECURE_REFERRER_POLICY default is 'same-origin' with versionchanged note; mention how to omit header (swev-id: django__django-12419).

rowan-stein committed:
Revert unintended global changes in docs/ref/settings.txt; will patch only SECURE_REFERRER_POLICY section (swev-id: django__django-12419).

rowan-stein committed:
Docs: SECURE_REFERRER_POLICY default updated to 'same-origin' with versionchanged note (swev-id: django__django-12419).

rowan-stein committed:
Reset docs/ref/settings.txt to base branch state to avoid duplication; will apply minimal patch (swev-id: django__django-12419).

rowan-stein committed:
Docs: Update SECURE_REFERRER_POLICY default to 'same-origin' and add versionchanged note (swev-id: django__django-12419).

rowan-stein committed:
Release notes: Document BC change for SECURE_REFERRER_POLICY default to 'same-origin' with mitigation guidance (swev-id: django__django-12419).

rowan-stein committed:
Docs: Mention default SECURE_REFERRER_POLICY is 'same-origin' in middleware docs (swev-id: django__django-12419).

rowan-stein commented:
Requesting review. Focus on:
- Correctness of default change in django/conf/global_settings.py.
- Documentation updates accuracy (ref/settings.txt, releases/3.1.txt, ref/middleware.txt) and BC note.
- Interactions with SecurityMiddleware behavior when SECURE_REFERRER_POLICY is None.
- Whether adding an explicit default-behavior test is desirable in this branch (optional).

noa-lucent commented on docs/releases/3.1.txt:
```diff
@@ -382,6 +382,8 @@ MariaDB 10.2 and higher.
 Miscellaneous
 -------------
 
+* Default Referrer-Policy changed: The default ``SECURE_REFERRER_POLICY`` is now ``'same-origin'`` (was ``None``). As a result, ``SecurityMiddleware`` will send a ``Referrer-Policy`` header by default. If your deployment relies on omitting or customizing this header, explicitly configure ``SECURE_REFERRER_POLICY`` appropriately (e.g., ``None`` to omit, or ``'no-referrer'``). Note: For HTTPS, Djan ...
```
[minor] Could we rewrap this bullet to Django's usual ~79 character width and split the mitigation guidance into separate sentences? The current single-line paragraph is very hard to read and doesn’t match the formatting of the surrounding release notes.

noa-lucent commented on docs/ref/middleware.txt:
```diff
@@ -296,7 +296,7 @@ for:
 the :setting:`SECURE_REFERRER_POLICY` setting (note spelling: browsers send a
 ``Referer`` header when a user clicks a link, but the header instructing a
-browser whether to do so is spelled ``Referrer-Policy``). The valid values for
+browser whether to do so is spelled ``Referrer-Policy``). By default, ``SECURE_REFERRER_POLICY`` is set to ``'same-origin'``, which applies referrers only to same-origin links. The valid values for
```
[minor] Could you wrap the new sentence to match the surrounding paragraph formatting? Right now the addition produces a single 200+ character line. Breaking before “By default…” (and wrapping subsequent text to ~79 chars) will keep the ReST source readable and consistent.

noa-lucent commented on docs/ref/settings.txt:
```diff
@@ -2395,7 +2395,10 @@ from URL paths, so patterns shouldn't include them, e.g.
+Default: ``'same-origin'``
+
+.. versionchanged:: 3.1
+   Default changed from ``None`` to ``'same-origin'``. Set ``SECURE_REFERRER_POLICY=None`` to omit the header.
```
[minor] Could we wrap this directive content and keep the inline code consistent with the rest of the docs? e.g.\n\n.. versionchanged:: 3.1\n   Default changed from None to 'same-origin'.\n   Set SECURE_REFERRER_POLICY = None to omit the header.\n\nThat keeps the lines under 80 chars and matches the spacing style used elsewhere.

noa-lucent commented on django/conf/global_settings.py:
```diff
@@ -637,6 +637,10 @@ def gettext_noop(s):
+#
+# Version changed: 3.1 — default changed from None to 'same-origin'.
+# Set SECURE_REFERRER_POLICY = None to omit the header entirely.
+SECURE_REFERRER_POLICY = 'same-origin'
```
[major] Could we add a regression test that exercises the new default? Right now all of the SecurityMiddleware tests override SECURE_REFERRER_POLICY. A test that leaves the default in place and asserts that the middleware emits Referrer-Policy: same-origin would lock in this behavior and guard against future regressions.

noa-lucent reviewed (changes requested):
Thanks for the update—this is a sensible default shift and SecurityMiddleware still omits the header when SECURE_REFERRER_POLICY is explicitly set to None.\n\nI’ve left a few notes: we should add a regression test that exercises the new default header value, and the documentation changes need a quick rewrap to match Django’s formatting conventions.

casey-brooks committed:
test(middleware): cover default referrer policy

Add regression coverage for SecurityMiddleware's default Referrer-Policy header and wrap related documentation to Django's formatting width.

casey-brooks commented:
### Tests & Lint
- `PYTHONPATH=/workspace/django python runtests.py middleware.test_security --parallel=1` (from `tests/`) — passed (24 tests, 0 failed, 0 skipped)
- `flake8 tests/middleware/test_security.py` — lint passed with no issues

noa-lucent reviewed (approved):
Thanks for addressing the feedback.\n\n• test_referrer_policy_default_same_origin now locks in the new default header, and existing overrides still behave correctly.\n• Release notes and setting docs are wrapped consistently, and the middleware docs clearly explain the default.\n\nLooks good to me.
